### PR TITLE
Redesign tab bar

### LINKED ISSUE #2563: Redesign tab bar
__What should we do?__

Adjust the tab bar to the [new design](https://github.com/camunda/camunda-modeler-design/issues/37).

image

image

image

image

[Prototype](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v17/)
[Kickoff Miro board](https://miro.com/app/board/o9J_lh4pDM0=/)

__Tasks__

* [ ] ensure all tab states
  * [ ] active tab shows the currently opened tab including the close icon
  * [ ] non-active tabs are slightly greyed-out
  * [ ] dirty marker is still present
* [ ] adjust tabs design
  * [ ] display icon of the entity (BPMN, DMN, Form, fallback?)
  * [ ] display title
  * [ ] display path as title tag
  * [ ] all tabs have equal sizes
  * [ ] keep space for tab actions
* [ ] ensure tabs behavior
  * [ ] it is possible to close all tabs, including welcome tab
  * [ ] tabs can be drag & dropped (keep the behavior)
  * [ ] new tabs are opened to the right of the active tab (keep the behavior)
  * [ ] the close button is only available on the active tab
  * [ ] make sure context menu on right click still exists and works properly
* [ ] ensure tabs scalability
  * [ ] remove horizontal scrolling entirely
  * [ ] tabs became smaller when they reach the available width of the tab area

__Why should we do it?__

Unified look and feel, and a user-friendly way to work with tabs.

---

Cf. [design issue](https://github.com/camunda/camunda-modeler-design/issues/37)
Child of https://github.com/camunda/camunda-modeler/issues/2461

### PR DESCRIPTION
**Pull request by pinussilvestrus**
Closes #2563 

**What's inside**

* add `getIcon` handler to tab providers
* redesign tab bar (cf. https://github.com/camunda/camunda-modeler/issues/2563)
* updated unsaved state

image

image

image

Find artifacts here to check it out (cf. [Build](https://github.com/camunda/camunda-modeler/actions/runs/1506683299))

* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-linux-x64.tar.gz
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-mac.dmg
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-mac.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-win-ia32.zip
* https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2563-redesign-tab-bar/camunda-modeler-2563-redesign-tab-bar-win-x64.zip

### CODE REVIEW

**Comment by pinussilvestrus** [comment]
I decided to keep the `TabLinks` inside the `primitives` to avoid bigger re-arrangements when not needed (although the TabLinks component is less primitive than before).

**Comment by barmac** [comment]
Found a glitch for a long name: 
image

**Comment by barmac** [comment]
Is this part of the design to display the dirty marker only for the active tab?

**Comment by pinussilvestrus** [comment]
> The "X" button is overflowing so it looks like an arrow. When I click the dirty marker, it triggers the close action.

Seems like I broke the existing dirty marker styles with this one. That shall be fixed, thanks!

> Is this part of the design to display the dirty marker only for the active tab?

The way the dirty marker is currently implemented is heavily dependant on the close handle. Since the close handle is not present on non-active tabs, the dirty marker also won't show up. We will update the whole dirty marker topic in a follow-up, that's why I completely left it untouched.

**Comment by barmac** [comment]
@andreasgeier The contrast here is super small:
image

**Review by barmac — CHANGES_REQUESTED** [review]
Generally, it looks good. There are a few issues to address before we merge this.

File: client/src/app/primitives/Tabbed.less
Comment (pinussilvestrus) [reply client/src/app/primitives/Tabbed.less:117]: That actually causes the bug described in https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-979818838 👍

**Comment by andreasgeier** [comment]
@pinussilvestrus 

I adjusted the not-active state to consider the feedback from @barmac regarding contrast. See updated [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/) for the new colors. It's the same color like the active state, I limited the differentiation to the underline for active state.

Screen Shot 2021-11-26 at 18 02 32

Also, please correct the style of the active state line. I made a CSS mistake in the prototype and the line got truncated by other lines. Again, see [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/) for the fix. This is the expected style (2px height, overlap other lines):

Screen Shot 2021-11-26 at 18 01 47

In the meantime, I added the unsaved marker as well. See [tab bar design issue](https://github.com/camunda/camunda-modeler-design/issues/37) and [prototype v16](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v16/).

**Comment by pinussilvestrus** [comment]
Markes as [WIP] to add requested changes
* add the new dirty marker
* adjust active and hover bottom line (2px height, overlay other lines)
* react to tiny state

image

**Comment by pinussilvestrus** [comment]
Regarding this one 

> This results in the tab line being properly set, but don't move when the window got resized (cf. screencast).

We can potentially re-render the lines once the window get resized (sketched via 077f5c4bef46c76b217f5aed3e361905b5f6815d).

Kapture 2021-11-30 at 15 57 37

Let's discuss this one in an orderly manner later this week @barmac @andreasgeier.

**Comment by pinussilvestrus** [comment]
Thanks to @andreasgeier we were able to resolve the tab bottom line with CSS only, no fixed positioning is needed 👍 

This is up for review now, the artifacts are up-to-date.

**Review by nikku — COMMENTED** [review]
It seems we accidentally broke re-arranging the tabs. Can you have a look?

As you can see in the capture below it does work, sometimes, only.

capture XozHoQ_optimized

**Comment by nikku** [comment]
Other than that, works very nice!

**Comment by pinussilvestrus** [comment]
Thanks for the report for drag & drop! That should be resolved now, artefacts are updated.

**Comment by barmac** [comment]
I am in the review right now (commit `a432241`).

**Comment by barmac** [comment]
https://user-images.githubusercontent.com/28307541/144406925-3d7579d5-fedb-45ba-889c-26942b28bd4d.mov

I have to change the tab to trigger change from the small tab to normal tab appearance.

File: client/src/app/primitives/TabLinks.js
Comment (barmac) [inline client/src/app/primitives/TabLinks.js:168]: Solution sketch for https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-984508557

```suggestion

  const tabNode = tabRef.current;

  // set tab to <small> once a defined threshold is reached (be responsive)
  React.useEffect(() => {
    if (!tabNode) {
      return;
    }

    const resizeObserver = new ResizeObserver(() => {
      const width = tabNode.getBoundingClientRect().width;

      tabNode.classList.toggle(TAB_SMALL_SELECTOR, width < SMALL_TAB_WIDTH);
    });

    resizeObserver.observe(tabNode);

    return () => resizeObserver.disconnect();
  }, [ tabNode ]);
```

Comment (barmac) [reply client/src/app/primitives/TabLinks.js:168]: So instead of checking the width on render, we do this whenever node is resized.

**Comment by barmac** [comment]
> It seems we accidentally broke re-arranging the tabs. Can you have a look?

This works for me now :)

File: client/src/app/primitives/TabLinks.js
Comment (pinussilvestrus) [reply client/src/app/primitives/TabLinks.js:168]: I think was is missing here is to react to the _initial size_ of a tab (first render). E.g. when I open 15+ tabs at once, it wouldn't react because there is no resizing included.

Comment (pinussilvestrus) [reply client/src/app/primitives/TabLinks.js:168]: Okay, actually the initial state would be respected as well. Instead, I'd probably handle the class via state, as we do with `dirty` and `active`. The reason is that it would override the classes on every render. Something like

```js
const [ small, setSmall ] = React.useState(false);

const tabNode = tabRef.current;

React.useEffect(() => {
  if (!tabNode) {
    return;
  }

  const resizeObserver = new ResizeObserver(() => {
    const width = tabNode.getBoundingClientRect().width;
    setSmall(width < SMALL_TAB_WIDTH);
  });

  resizeObserver.observe(tabNode);
  return () => resizeObserver.disconnect();
}, [ tabNode ]);

return (
  <div
    className={ classNames('tab', {
      'tab--active': active,
      'tab--dirty': dirty,
      'tab--small': small
      }) } >
      
      // ...
  </div>
); 
```

What do you think?

Comment (barmac) [reply client/src/app/primitives/TabLinks.js:168]: LGTM

**Comment by pinussilvestrus** [comment]
I pushed the resize observer behavior on top as suggested in https://github.com/camunda/camunda-modeler/pull/2589#discussion_r760983981. I don't build new artifacts for this but wait for design review first.

**Review by barmac — APPROVED** [review]

**Review by nikku — APPROVED** [review]

**Comment by pinussilvestrus** [comment]
@andreasgeier will you review this one in particular or shall we wait once the tab action is added for an overall design review?

**Comment by andreasgeier** [comment]
@pinussilvestrus I'm on it.

- entity icons are blurred because they are scaled. please ensure using a 100% display with sharp lines.
- font size should be 14px
- dirty marker should be 8x8px and should have a border in dark blue
- close icon looks scaled

please adjust the overall style to match the proto.

proto vs implementation:

Screen Shot 2021-12-03 at 15 40 01

**Comment by andreasgeier** [comment]
The tab bar looks aery in the prototype because it is part of the unified header design. In the current implementation, it might look compressed because it is docked to the window frame. I think we should fix this by using the same height as the status bar (32px) but I can't judge it with the current styling issues.

We could try this out at a later state, I just wanted to mention it if you could easily try out and paste a screenshot.

**Comment by pinussilvestrus** [comment]
Thanks @andreasgeier, I will follow up with the tiny fixes next week 👍

**Comment by andreasgeier** [comment]
To distinguish between the different states, I would like to propose these title tags:

Newly created file, not yet saved: `(new file) – unsaved`
Existing file, untouched: `path/to/name.bpmn`
Existing file in dirty state: `path/to/name.bpmn – unsaved`

**Comment by pinussilvestrus** [comment]
@andreasgeier I adjusted the design accordingly (unscaled icons, font-size 14px, dirty marker 8x8px). The titles are also updated to the one [you suggested](https://github.com/camunda/camunda-modeler/pull/2589#issuecomment-985592073). Please note that there is currently bug related to unsaved --> then saved tabs, which is not related to this PR (cf. https://github.com/camunda/camunda-modeler/issues/2581). The artefacts are also updated.

I also increased the size of the tab bar to the same as the status bar. See the following screenshot:

Bildschirmfoto 2021-12-05 um 20 40 35

**Comment by andreasgeier** [comment]
As discussed, please adjust the entity icon (and in relation the dirty marker) to be centered between divider and tab title.
Everything else looks as expected. Nice Work!

✅ &nbsp; Approved! Pls merge manually after the fix 🙂

**Comment by pinussilvestrus** [comment]
Last styling issues are pushed, let's merge this one 👍

**Pull request merged**
